Framework/Algorithms/src/AverageSpectrumBackground.cpp

```diff
@@ -0,0 +1,151 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2019 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+
+#include "MantidAlgorithms/AverageSpectrumBackground.h"
+#include "MantidAPI/CommonBinsValidator.h"
+#include "MantidKernel/ArrayLengthValidator.h"
+#include "MantidKernel/ArrayProperty.h"
+
+namespace Mantid {
+namespace Algorithms {
+
+/// Algorithm's name for identification. @see Algorithm::name
+const std::string AverageSpectrumBackground::name() const {
+  return "AverageSpectrumBackground";
+}
+
+/// Algorithm's version for identification. @see Algorithm::version
+int AverageSpectrumBackground::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+const std::string AverageSpectrumBackground::category() const {
+  return "Reflectometry;Reflectometry\\ISIS";
+}
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string AverageSpectrumBackground::summary() const { return ""; }
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(AverageSpectrumBackground)
+
+/** Sums spectra bin by bin in the given range using the child algorithm
+ * GroupDetectors. Note that KeepUngroupedSpectra is set to false.
+ * @param inputWS:: The Workspace2D to take as input
+ * @param indexList:: An array of workspace indices to combine
+ * @return :: A workspace containing a single spectra and DectectorGroups
+ */
+API::MatrixWorkspace_sptr AverageSpectrumBackground::groupBackgroundDetectors(
+    API::MatrixWorkspace_sptr inputWS, const std::vector<size_t> indexList) {
+
+  auto alg = this->createChildAlgorithm("GroupDetectors");
+  alg->setProperty("InputWorkspace", inputWS);
+  alg->setProperty("WorkspaceIndexList", indexList);
+  alg->setProperty("KeepUngroupedSpectra", false);
+  alg->execute();
+  return alg->getProperty("OutputWorkspace");
+}
+
+std::vector<size_t> AverageSpectrumBackground::getSpectraFromRange(
+    const std::vector<size_t> range) {
+  const auto start = range[0];
+  const auto end = range[1];
+  std::vector<size_t> spectra;
+  for (auto itr = start; itr < end; ++itr) {
+    spectra.push_back(itr);
+  }
+  return spectra;
+}
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void AverageSpectrumBackground::init() {
+
+  // Input workspace
+  declareProperty(
+      Kernel::make_unique<API::WorkspaceProperty<API::MatrixWorkspace>>(
+          "InputWorkspace", "", Kernel::Direction::Input,
+          boost::make_shared<API::CommonBinsValidator>()),
+      "An input workspace.");
+
+  // bottom background range
+  declareProperty(Kernel::make_unique<Kernel::ArrayProperty<size_t>>(
```

Comment:
Could we use [IndexProperty](http://docs.mantidproject.org/v3.11.0/concepts/IndexProperty.html) here?

One thing to note is that we recently changed the `ProcessingInstructions` to `RRO` to take spectrum numbers instead of workspace indices because that is easier for the users to deal with. I expect they would want to use spectrum numbers here as well (IndexProperty would allow them to do both). We would have liked to use IndexProperty in `RRO` but this was not possible due to requiring multiple output groups (see #14000 for details), but that probably isn't a limitation here because they get summed anyway into a single spectrum anyway?